xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java

```diff
@@ -921,17 +912,29 @@ public Collection<String> getComponentsDomains() {
 
     @Override
     public boolean hasClientRoute(JID jid) {
-        return usersCache.containsKey(jid.toString()) || isAnonymousRoute(jid);
+        final Lock lock = usersSessionsCache.getLock(jid.toBareJID());
+        lock.lock();
+        try {
+            return usersCache.containsKey(jid.toFullJID()) || isAnonymousRoute(jid);
```

Comment:
I forget how these locks work (it's been a while) but I assume they are re-entrant?

This method 
* invokes `usersSessionsCache.getLock(jid.toBareJID());`
* then calls `isAnonymousRoute(jid)`
  * and this method also invokes `usersSessionsCache.getLock(jid.toBareJID());`

I'd expect any tests to throw up issues, but that bit worries me. It looks wrong on inspection. Perhaps you need a `private boolean isAnonymousRouteUnlocked(JID jid)` that doesn't take out a lock, that this method (and the public one) can take out knowing that it's already locked?

Reply:
Yes, the locks are re-entrant. I've checked that for both Openfire's own caches, as well as Hazelcast's. Your other comment fixed a typo in a comment that explains that for another case that uses pretty much the same approach. :)

If your concern is optics on inspection, then I'd rather fix that by adding additional code commenting. I think that leads to easier-to-maintain code.